kernel/syscall.c

```diff
@@ -3731,31 +3756,147 @@ static long _SYS_munmap(
         }
     }
 
+    myst_lockfs_lock();
+    myst_mman_lock();
+
     /* Handle MAP_SHARED unmapping */
     {
         bool is_shmem;
         if (myst_shmem_handle_munmap(addr, length, &is_shmem) < 0)
-            return (_return(n, -EINVAL));
+            ERAISE(-EINVAL);
 
         if (is_shmem)
-            return (_return(n, 0));
+            goto done;
     }
 
     if (!myst_process_owns_mem_range(addr, length, true))
-        return (_return(n, -EINVAL));
+        ERAISE(-EINVAL);
 
-    long ret = (long)myst_munmap(addr, length);
+    ret = (long)myst_munmap(addr, length);
 
     if (ret == 0)
     {
         /* set ownership this mapping to nobody */
         if (myst_mman_pids_set(addr, length, 0) != 0)
+        {
+            myst_mman_unlock();
+            myst_lockfs_unlock();
             myst_panic("myst_mman_pids_set()");
+        }
     }
 
+done:
+    myst_mman_unlock();
+    myst_lockfs_unlock();
+
     return (_return(n, ret));
 }
 
+static long _SYS_mremap(long n, long params[6])
+{
+    void* old_address = (void*)params[0];
+    size_t old_size = (size_t)params[1];
+    size_t new_size = (size_t)params[2];
```

Comment:
If ``new_size`` is zero then return ``-EINVAL``. Check range of values for ``new_size`` and ``old_size``.